gusto/limiters.py

```diff
@@ -0,0 +1,117 @@
+from __future__ import absolute_import, print_function, division
+from firedrake import dx
+from firedrake.function import Function
+from firedrake.functionspace import FunctionSpace
+from firedrake.parloops import par_loop, READ, RW, INC
+from firedrake.slope_limiter.vertex_based_limiter import VertexBasedLimiter
+
+__all__ = ["ThetaLimiter"]
+
+_copy_into_Q1DG_loop = """
+theta_hat[0][0] = theta[0][0];
+theta_hat[1][0] = theta[1][0];
+theta_hat[2][0] = theta[3][0];
+theta_hat[3][0] = theta[4][0];
+"""
+_copy_from_Q1DG_loop = """
+theta[0][0] = theta_hat[0][0];
+theta[1][0] = theta_hat[1][0];
+theta[3][0] = theta_hat[2][0];
+theta[4][0] = theta_hat[3][0];
+"""
+
+_check_midpoint_values_loop = """
+if (theta[2][0] > fmax(theta[0][0], theta[1][0]))
+    theta[2][0] = 0.5 * (theta[0][0] + theta[1][0]);
+else if (theta[2][0] < fmin(theta[0][0], theta[1][0]))
+    theta[2][0] = 0.5 * (theta[0][0] + theta[1][0]);
+if (theta[5][0] > fmax(theta[3][0], theta[4][0]))
+    theta[5][0] = 0.5 * (theta[3][0] + theta[4][0]);
+else if (theta[5][0] < fmin(theta[3][0], theta[4][0]))
+    theta[5][0] = 0.5 * (theta[3][0] + theta[4][0]);
+"""
+
+_weight_kernel = """
+for (int i=0; i<weight.dofs; ++i) {
+    weight[i][0] += 1.0;
+    }"""
+
+_average_kernel = """
+for (int i=0; i<vrec.dofs; ++i) {
+        vrec[i][0] += v_b[i][0]/weight[i][0];
+        }"""
+
+
+class ThetaLimiter(object):
+    """
+    A vertex based limiter for fields in the DG1xCG2 space,
+    i.e. temperature variables. This acts like the vertex-based
+    limiter implemented in Firedrake, but in addition corrects
+    the central nodes to prevent new maxima or minima forming.
+    """
+
+    def __init__(self, space):
+        """
+        Initialise limiter
+
+        :param space : FunctionSpace instance
+        """
+
+        self.Vt = space
+        self.Q1DG = FunctionSpace(self.Vt.mesh(), 'DG', 1)  # space with only vertex DOF
```

Comment:
Do you think we should check that the space is DG1xCG2?

Reply:
I'm not sure how I would do that!

Reply:
There's an `is_cg` test somewhere in `transport_equation.py` which I just copied from somewhere in the Firedrake source, which checks the number of DOFs is as expected - we could move that to a `utility_functions.py` file so we can use it in both places. I think @colinjcotter doesn't like it, but at least if we have it defined in just one place we can change it if we think of a better way to do it! You can get the degree of the space using `space.ufl_element().degree()`.

Reply:
There is an alternative way to check the element continuity. If `W` is your DG1xCG2 `MixedElement`, then `W.ufl_element().sub_elements()` gives you the individual component elements. Each of those elements has a `sobolev_space()` method which tells you the Sobolev space the element is associated with. That is, If you write `DG1.ufl_element().sobolev_space().name`, you'll get `L2`. Similarly, `CG2.ufl_element().sobolev_space().name` returns `H1`.

If `W` is a `TensorProductElement`, this is simplified a bit. Calling `sobolev_space()` on the TPE gives you a `DirectionalH` sobolev space, which you can pass a spatial index to see the continuity in a particular direction. For example, `W.sobolev_space()[i].name` returns `L2` for i a horizontal index (0, 1 in 3D), and the vertical index (2 in 3D) returns `H1`.

Just an alternative you may want to experiment with!